### PR TITLE
Avoid fail fetch when apiKey not defined

### PR DESCRIPTION
**Pull request by edellucien**
- Add apiKey parameter to MapboxLayer
 - Add apiKey parameter to Mapbox Map
 - Warning in console when apiKey is missing
 - Do not required apiKey parameter when it is set to false.

# Others

- [x] It's not a hack or at least an unauthorized hack :).
- [ ] The images added are optimized.
- [x] Everything in ticket description has been fixed.
- [x] The author of the MR has made its own review before assigning the reviewer.
- [x] The title means something for a human being.
- [x] The title contains [WIP] if it's necessary.
- [ ] Labels applied. if it's a release? a hotfix?
- [x] The new class' members & methods are well documented
- [x] Tests added.

### CODE REVIEW

File: src/common/utils/getUrlParams.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/**
```
Comment (mhaertwig) [inline src/common/utils/getUrlParams.js:1]: https://github.com/geops/mobility-toolbox-js/blob/master/package.json#L17

Please use `query-string`, which is already a dependency.

File: src/mapbox/Map.js [hunk 2]
```diff
@@ -14,6 +16,25 @@ class Map extends mixin(MBMap) {
...
+    const { style, apiKey } = options;
+    if (apiKey !== false) {
+      let newStyle;
+      const parameters = getUrlParams(style);
+
+      if (!apiKey) {
+        // eslint-disable-next-line no-console
+        console.warn(`No apiKey is defined for Mapbox Map request to ${style}`);
+      } else if (parameters.key) {
+        // replace key value from apiKey parameter.
```
Comment (mhaertwig) [inline src/mapbox/Map.js:28]: Use `query-string`

File: src/mapbox/Map.js [hunk 3]
```diff
@@ -14,6 +16,25 @@ class Map extends mixin(MBMap) {
...
+        parameters.key = apiKey;
+        const [baseUrl] = style.split('?');
+        newStyle = `${baseUrl}?${qs.stringify(parameters)}`;
+      } else {
+        newStyle = `${style}?key=${apiKey}`;
+      }
+      // eslint-disable-next-line no-param-reassign
+      options.style = newStyle;
```
Comment (mhaertwig) [inline src/mapbox/Map.js:36]: Please find a better solution without `// eslint-disable-something`

**Review by mhaertwig — CHANGES_REQUESTED** [review]
Good, but could be simpler and more robust.

**Review by mhaertwig — APPROVED** [review]

**Comment by oterral** [comment]
@mhaertwig  Are you sure you want this ?   It blocks the possibility to have an external mapbox layer as layer.

File: src/mapbox/Map.js [hunk 4]
```diff
@@ -14,10 +15,27 @@ class Map extends mixin(MBMap) {
    * Constructor.
    */
   constructor(options) {
+    const { style, apiKey } = options;
+    let newStyle = options.style;
+    if (apiKey !== false) {
```
Comment (oterral) [inline src/mapbox/Map.js:20]: apiKey should  be false by default.

File: src/mapbox/Map.js [hunk 5]
```diff
@@ -14,10 +15,27 @@ class Map extends mixin(MBMap) {
...
+        // eslint-disable-next-line no-console
+        console.warn(`No apiKey is defined for Mapbox Map request to ${style}`);
+      } else if (parameters.key) {
+        // replace key value from apiKey parameter.
+        parameters.key = apiKey;
+        newStyle = `${baseUrl}?${qs.stringify(parameters)}`;
+      } else {
+        newStyle = `${style}?key=${apiKey}`;
+      }
+    }
```
Comment (oterral) [inline src/mapbox/Map.js:33]: Use parseUrl and stringifyUrl . it should simplify the code

File: src/ol/layers/MapboxLayer.js [hunk 6]
```diff
@@ -268,6 +315,9 @@ export default class MapboxLayer extends Layer {
       }
     }
     this.mbMap.on('idle', this.updateAttribution);
+    // returns value to allow testing easily.
+    // eslint-disable-next-line consistent-return
+    return style;
```
Comment (oterral) [inline src/ol/layers/MapboxLayer.js:320]: Remove this.  Test the content of this.mbMap . mapboxgl.Map is already   mocked somewhere.

File: src/ol/layers/MapboxLayer.js [hunk 7]
```diff
@@ -173,6 +191,34 @@ export default class MapboxLayer extends Layer {
...
+      if (parameters[this.apiKeyName]) {
+        // replace key value from apiKey parameter.
+        parameters[this.apiKeyName] = this.apiKey;
+        style = `${baseUrl}?${qs.stringify(parameters)}`;
+      } else {
+        style = `${this.styleUrl}?${this.apiKeyName}=${this.apiKey}`;
+      }
+    }
+    return style;
+  }
```
Comment (oterral) [inline src/ol/layers/MapboxLayer.js:220]: looks like there is a duplication of code.

File: src/ol/layers/MapboxStyleLayer.test.js [hunk 8]
```diff
@@ -18,6 +18,7 @@ describe('MapboxStyleLayer', () => {
     onClick = jest.fn();
     source = new MapboxLayer({
       name: 'Layer',
+      apiKey: false,
```
Comment (oterral) [inline src/ol/layers/MapboxStyleLayer.test.js:21]: must work without specifiy apiKey

File: src/ol/layers/MapboxLayer.js (hunk 7 above)
Comment (edellucien) [reply src/ol/layers/MapboxLayer.js]: Centralized in a util fct 'getMapboxStyle'

**Comment by edellucien** [comment]
Thanks for the observations! I updated the code.

File: src/common/utils/getMapboxStyle.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        query: parsedStyle.query,
+      });
+    } else {
+      style = qs.stringifyUrl({
+        url: styleUrl,
+        query: {
+          [apiKeyName]: apiKey,
+        },
+      });
+    }
```
Comment (oterral) [inline src/common/utils/getMapboxStyle.js:33]: Make it simple:
```suggestion
      return  qs.stringifyUrl({
      ...parsedStyle
        query: {
          ...parsedStyle.query,
          [apiKeyName]: apiKey,
        },
      });
```

File: src/common/utils/getMapboxStyle.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Return the copyright a Mapbox map.
+ * @param {mapboxgl.Map} map A Mapbox map
+ * @ignore
+ */
+const getMapboxStyle = (apiKey, apiKeyName, styleUrl) => {
+  let style;
+  if (apiKey === false) {
+    style = styleUrl;
```
Comment (oterral) [inline src/common/utils/getMapboxStyle.js:11]:
```suggestion
    return styleUrl;
```

File: src/common/utils/getMapboxStyle.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  } else {
+    const parsedStyle = qs.parseUrl(styleUrl);
+    if (!apiKey) {
```
Comment (oterral) [inline src/common/utils/getMapboxStyle.js:14]: I see an other case.
```
if (!apiKey && parsedStyle.query[apiKeyName]) {
  return styleUrl;
}

if (!apiKey) {
  console.warn(`No apiKey is defined for request to ${styleUrl}`);
￼  return null;
}
```

File: src/mapbox/Map.js [hunk 12]
```diff
@@ -14,10 +15,16 @@ class Map extends mixin(MBMap) {
    * Constructor.
    */
   constructor(options) {
+    const { style } = options;
+    const apiKey = options.apiKey || false;
+    const apiKeyName = options.apiKeyName || 'key';
+    let newStyle = options.style;
+    newStyle = getMapboxStyle(apiKey, apiKeyName, style);
```
Comment (oterral) [inline src/mapbox/Map.js:22]: In getMapboxStyle you should manage the case where options.style is already a mapbox style object  and not a url

File: src/ol/layers/MapboxLayer.js [hunk 13]
```diff
@@ -173,6 +191,13 @@ export default class MapboxLayer extends Layer {
     );
   }
 
+  /**
+   * Returns a style URL with apiKey & apiKeyName infos.
+   */
+  createStyleUrl() {
```
Comment (oterral) [inline src/ol/layers/MapboxLayer.js:197]: useless function , use directly getMapboxStyle

**Review by oterral — CHANGES_REQUESTED** [review]
Some more use cases to manage

File: src/mapbox/Map.js (hunk 12 above)
Comment (edellucien) [reply src/mapbox/Map.js:22]: The way it is used in the constructor don't prevent this to happen actually?

File: src/ol/layers/MapboxLayer.js (hunk 13 above)
Comment (edellucien) [reply src/ol/layers/MapboxLayer.js]: We actually need to be able to call this function in target app so we keep this fct.

**Comment by edellucien** [comment]
> Some more use cases to manage

All use cases mentioned are taken into account :+1:

File: src/common/utils/getMapboxStyleUrl.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import qs from 'query-string';
+
+/**
+ * Return the copyright a Mapbox map.
+ * @param {mapboxgl.Map} map A Mapbox map
+ * @ignore
```
Comment (oterral) [inline src/common/utils/getMapboxStyleUrl.js:6]: change the function description

**Review by oterral — APPROVED** [review]
you can merge

**Pull request merged**